tests/template_tests/filter_tests/test_truncatechars_html.py

```diff
@@ -41,7 +41,7 @@ def test_truncate_unicode(self):
         )
 
     def test_truncate_something(self):
-        self.assertEqual(truncatechars_html("a<b>b</b>c", 3), "a<b>b</b>c")
+        self.assertEqual(truncatechars_html("a<b>b</b>c", 3), "a<b>b</b>…")
```

Comment:
Hmm... This looks wrong. We should return the full output if the length is exact. Maybe we've got an off-by-one issue somewhere?

Reply:
I'm not quite sure of the correct wording to use here. 

I think this is some kind of look ahead issue where we have reached the required length and truncate it. At this point there may/may not be more text remaining in the data. Maybe we could look at the parser's `rawdata` and parse that to see if there's anything left. I'm not sure that additional complexity is worth it. We've already added a note to say there will be small changes in the output.

Reply:
Ok, so this is definitely a regression we're introducing here that we should fix.

I added the following test to `tests/template_tests/filter_tests/test_truncatechars.py` which passes fine:

```python
    @setup({"truncatechars04": "{{ a|truncatechars:3 }}"})
    def test_truncatechars01(self):
        output = self.engine.render_to_string(
            "truncatechars04", {"a": "abc"}
        )
        self.assertEqual(output, "abc")
```

I think we should add this as it covers the exact length case which is missing there.

The HTML chars case is regressing. The problem is on this line:

https://github.com/django/django/pull/16421/files#diff-a6acaa0a744b3a7c841d9ab3ccbc1765f5749e3782fced61edced456c227b43fR188

We are passing in `length=truncate_len` when we should be passing in `length=length`, as for the words case. When you look at the text chars case we are passing through both `length` and `truncate_len`.

I suspect that we will want to hoist the following into a global helper:

https://github.com/django/django/pull/16421/files#diff-a6acaa0a744b3a7c841d9ab3ccbc1765f5749e3782fced61edced456c227b43fR180-R186

Which we can then call independently from the appropriate place in the text branch or the HTML branch (inside of `TruncateCharsHTMLParser`).